zkml/src/layers/transformer/layernorm.rs

```diff
@@ -1,33 +1,260 @@
+        self.gamma.shape[0]
+    }
+
+    /// Returns the [`QuantisedLayerNormData`] if there is any.
+    pub fn quant_info(&self) -> Option<&QuantisedLayerNormData> {
+        self.quant_info.as_ref()
+    }
+
+    /// Quantise the layer. To do this we want to have a common scale factor so that lookup tables can be reused, so we use the
+    /// constant [`LAYERNORM_SCALE_FACTOR`] as the input column scale factor. We need to work out how big the table needs to be to cover
+    /// all of our possible inputs.
+    ///
+    /// This method reutnrs the quantised [`LayerNorm`] as well as the `intermediate_bit_size` for the following requant layer.
+    pub fn quantise(
+        &self,
+        input_scaling: ScalingFactor,
+        model_scaling: ScalingFactor,
+    ) -> Result<(LayerNorm<Element>, usize, ScalingFactor)> {
+        // The input to the lookup table is `N*sum2 - sum1^{2}` where `sum2 = \sum xi^{2}` and `sum1 = \sum xi`.
+        // We use this value because the standard deviation can be calculated by `(N*sum2 - sum1^{2}).sqrt() / N`
+        // Since each `xi` is a value between `*quantisation::MIN` and `*quantisation::MAX` it has bit-size `*quantization::BIT_LEN - 1`.
+        // This means `sum1` has bit-size `ceil_log2(N) + *quantization::BIT_LEN - 1` and `sum2` has bit-size `2(*quantization::BIT_LEN - 1)`
+        // Then `sum1^{2}` has bit-size `2(ceil_log2(N) + *quantization::BIT_LEN - 1)` and `Nsum2` has bit_size `ceil_log2(N) + 2(*quantization::BIT_LEN - 1)`.
+        // Finally we have to multiply all of this by `multiplier = LAYERNORM_SCALE_FACTOR * input_scaling.scale() * input_scaling.scale()` so we have `ceil_log2(multiplier)`
+        // additional bits on top of this.
+
+        // Get the input scale
+        let input_scale = input_scaling.scale();
+        // Get the dim size (N)
+        let dim_size = self.normalisation_dim_size();
+        // We work out what we have to mutliply by so that everything is scaled to `LAYERNORM_SCALE_FACTOR` in quantised world
+        let multiplier =
+            (LAYERNORM_SCALE_FACTOR as f32 * input_scale * input_scale).round() as Element;
+        // Work out the number of variables the table requires, this is likely to be far too large to actually materialise as a table
+        let full_table_bit_size = 2 * (ceil_log2(dim_size) + *quantization::BIT_LEN - 1)
+            + ceil_log2(multiplier as usize)
+            + 1;
+        // To get around this we use the fact that we should only have roughly `2*(*quantization::BIT_LEN -1)` bits of precision i.e. only the most significant `2*(*quantization::BIT_LEN -1)`
+        // can actually be "trusted" the rest are essentially junk because they don't come from the actual inputs and are just guesses at the part that we have alread "rounded away" in quantisation.
+        // So the actual part we perform inverse square root on is size `2*(*quantization::BIT_LEN -1)` and then we rust need the discarded part to be range checked.
```

Comment:
> then we rust need the discarded part to be range checked.
 
 can you reformulate that part please ? if we need to range check it then we should and there is no trust in that. If we don't then  let's just say we don't.